sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -2634,13 +2651,21 @@ pi_result piQueueRetain(pi_queue Queue) {
 }
 
 pi_result piQueueRelease(pi_queue Queue) {
+  return QueueRelease(Queue, nullptr);
+}
+
+static pi_result QueueRelease(pi_queue Queue, pi_queue LockedQueue) {
   PI_ASSERT(Queue, PI_INVALID_QUEUE);
   // We need to use a bool variable here to check the condition that
   // RefCount becomes zero atomically with PiQueueMutex lock.
   // Then, we can release the lock before we remove the Queue below.
   bool RefCountZero = false;
   {
-    std::lock_guard<std::mutex> Lock(Queue->PiQueueMutex);
+    std::mutex NoLock;
+    auto Lock = ((Queue == LockedQueue)
```

Comment:
Thank you, approach looks good to me.

Just one thing, it looks better to use unique_lock when we have a condition, because unique_lock can be created with no associated mutex:

```
    auto Lock = ((Queue == LockedQueue)
                           std::unique_lock<std::mutex>() 
                         : std::unique_lock<std::mutex>(Queue->PiQueueMutex));
```

Reply:
But with unique_lock you have to lock/unlock manually which breaks RAII

Reply:
No, it has the same semantics as lock_guard. Constructor locks the mutex and destructor unlocks it.

Reply:
It has different semantics only if it is created with some of this: https://en.cppreference.com/w/cpp/thread/lock_tag_t

if lock tag is not provided then mutex is locked automatically just like with lock_guard.

Reply:
OK then, I will change to unique_lock.